contracts/subdomainregistrar/BaseSubdomainRegistrar.sol

```diff
@@ -0,0 +1,263 @@
+
+        (address token, uint256 fee) = ISubdomainPricer(
+            names[parentNode].pricer
+        ).price(parentNode, label, duration);
+
+        _checkParent(parentNode, duration);
+
+        if (fee > 0) {
+            IERC20(token).transferFrom(
+                msg.sender,
+                address(names[parentNode].beneficiary),
+                fee
+            );
+        }
+
+        _register(
+            parentNode,
+            label,
+            newOwner,
+            resolver,
+            fuses,
+            uint64(block.timestamp) + duration,
+            records
+        );
+    }
+
+    /* Internal Functions */
+
+    function _register(
+        bytes32 parentNode,
+        string calldata label,
+        address newOwner,
+        address resolver,
+        uint32 fuses,
+        uint64 expiry,
+        bytes[] calldata records
+    ) internal {
+        bytes32 node = keccak256(
+            abi.encodePacked(parentNode, keccak256(bytes(label)))
+        );
+
+        if (!available(node)) {
+            revert Unavailable();
+        }
+
+        if (records.length > 0) {
+            wrapper.setSubnodeOwner(
+                parentNode,
+                label,
+                address(this),
+                0,
+                expiry
+            );
+            _setRecords(node, resolver, records);
+        }
+
+        wrapper.setSubnodeRecord(
+            parentNode,
+            label,
+            newOwner,
+            resolver,
+            0,
+            fuses | PARENT_CANNOT_CONTROL, // burn the ability for the parent to control
+            expiry
+        );
+
+        emit NameRegistered(node, expiry);
+    }
+
+    function _batchPayBeneficiary(
+        bytes32 parentNode,
+        string[] calldata labels,
+        uint64 duration
+    ) internal {
+        ISubdomainPricer pricer = names[parentNode].pricer;
+        for (uint256 i = 0; i < labels.length; i++) {
+            (address token, uint256 price) = pricer.price(
+                parentNode,
+                labels[i],
+                duration
+            );
+            IERC20(token).transferFrom(
+                msg.sender,
+                names[parentNode].beneficiary,
+                price
+            );
+        }
+    }
+
+    function _setRecords(
+        bytes32 node,
+        address resolver,
+        bytes[] calldata records
+    ) internal {
+        for (uint256 i = 0; i < records.length; i++) {
+            // check first few bytes are namehash
+            bytes32 txNamehash = bytes32(records[i][4:36]);
+            require(
+                txNamehash == node,
+                "SubdomainRegistrar: Namehash on record do not match the name being registered"
+            );
+            resolver.functionCall(
+                records[i],
+                "SubdomainRegistrar: Failed to set Record"
+            );
+        }
+    }
+
+    function _checkParent(bytes32 parentNode, uint64 duration) internal view {
```

Comment:
I don't think test coverage exists for these cases

- parent name is not wrapped -> Error
- parent is NOT .eth (DNS, subname)
- parent expiry is in grace period
- parent expiry does not exist (subname?)